### PR TITLE
Fix exception handling for overloaded operators.

### PR DESCRIPTION
**Pull request by insertinterestingnamehere**
Initial version that only handles binary operators. I'm still working on the code for unary operators. Tests pending.

### CODE REVIEW

**Comment by insertinterestingnamehere** [comment]
Here's a status update:
This is mostly working. I still need to fix the indexing operator and have uncovered more work to be done on assignment. There seem to be some issues currently present in overloading the ! and & operators. the ! operator doesn't always parse correctly when it is used, and the & operator doesn't follow the type declaration given in the overload if it differs from just taking the address. I'll look at fixing those, but that is a somewhat lower priority and may be a better topic for a different PR anyway.

Out of curiosity, what is the difference between `calculate_result_code` and `generate_result_code`? The two appear to do essentially the same thing. I've applied these changes to `generate_result code` and it is working well. When do I need to worry about `calculate_result_code`?

**Comment by insertinterestingnamehere** [comment]
A quick note about design. In order to make sure that operators using one exception handler can properly raise exceptions when used as a subexpression of another operation that uses a different exception handler, I've made it so that each node corresponding to an operation where an exception might be raised marks itself as temporary.
This works well in all cases *except* when an operator returns a reference but is not used as an lvalue (as can happen with indexing). In that case, the temporary name should still be set aside for use, but it should not be declared until it is initialized. I'm currently trying to determine the best way to make that happen.

**Comment by insertinterestingnamehere** [comment]
Okay, this is ready. All currently overloadable operators now obey the proper c++ exception catching semantics. In-place operators, `operator||`, and `operator&&` still aren't enabled, but I'm out of time to work on that at the moment.

In order to account for the fact that exception handlers may be different for different expressions, I had to break each subexpression into its own try-except block. That worked in nearly every case.

One particularly special case is when a user is simultaneously using an overloaded `operator[]` and an overloaded `operator=`. In that case, since it is impossible to declare an uninitialized lvalue reference, and try-catch statements have their own scope, I had to allow for the different exception handling routines via a nested try-except statement. The inner block catches errors from indexing and the outer block catches errors from assignment. The only incongruency with the standard rules is that if an exception is uncaught by the exception handling routine for indexing, it can still be caught by the exception handling routine for assignment. The only alternative was to force users to use the same exception handlers for both operators, and I figured allowing the layered exception handling was the better option.

**Comment by insertinterestingnamehere** [comment]
Quick correction to the previous summary: exceptions not caught by the overloaded assignment operator can be caught by the overloaded indexing operator, not vice versa.

**Comment by insertinterestingnamehere** [comment]
The structure looks roughly like this:

    try:
        (make a temporary lvalue reference from indexing)
        try:
            (assign to the reference)
        catch:
            (call the assignment exception handler)
    catch:
        (call the indexing exception handler)

**Comment by insertinterestingnamehere** [comment]
Any chance this could still be included in the 0.23 release?

**Comment by insertinterestingnamehere** [comment]
Okay, I've just updated this to make sure temporaries are allocated when handling exceptions from function calls and `operator bool` as well.

File: tests/run/cpp_operator_exc_handling.srctree [hunk 1]
```diff
@@ -0,0 +1,485 @@
...
+PYTHON -c "from check_operator_exception_handling import test; test()"
+"""
+
+######## setup.py ########
+
+from distutils.core import setup
+from Cython.Build import cythonize
+setup(ext_modules=cythonize('*.pyx', language='c++'))
+
+######## wint.cpp ########
```
Comment (robertwb) [inline tests/run/cpp_operator_exc_handling.srctree]: No need for this to be a srctree test, could you instead make it a standard test (if you have foo.pyx, any foo_helper.cpp will be available as well).

Comment (insertinterestingnamehere) [inline tests/run/cpp_operator_exc_handling.srctree]: Will do. I'll take care of that now.

**Comment by robertwb** [comment]
I haven't gone over all of it so far, but mostly looking good. Probably a bit of a big change to slip into 0.23 at this point, but I'd be happy to do a subsequent release shortly.

**Comment by insertinterestingnamehere** [comment]
Okay, sounds good to me. Thanks!

**Comment by insertinterestingnamehere** [comment]
Rebased.

**Comment by scoder** [comment]
Thanks! Looks good to me, definitely good enough for a merge. The indentation style is a bit non-PEP8-ish in some places (e.g. multi-line function signatures), but that can be fixed later - and I'd be happy to receive an update patch for this. :)

**Pull request merged**